### PR TITLE
CMake: option to build tests

### PR DESCRIPTION
**Pull request by mattkdab**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -219,34 +221,39 @@ endif()
...
+    add_test(NAME cxx_qt_gen_test_inputs_gen COMMAND rustfmt --check ${CXX_QT_GEN_TEST_INPUTS})
+    add_test(NAME cxx_qt_gen_test_outputs_gen COMMAND rustfmt --check ${CXX_QT_GEN_TEST_OUTPUTS})
+    
+    # Add test which checks that a build rerun doesn't recompile and uses caches instead
+    add_test(NAME cargo_build_rerun COMMAND "${CMAKE_CURRENT_SOURCE_DIR}/scripts/check_cargo_build_rerun.sh" "${CMAKE_CURRENT_SOURCE_DIR}")
+    
+    # Ensure that cargo_build_rerun doesn't run while we are already building
+    set_tests_properties(cargo_build_rerun PROPERTIES RUN_SERIAL TRUE)
+    
+    add_subdirectory(tests)
```
Comment (ahayzen-kdab) [inline CMakeLists.txt:255]: think this add_subdirectory might need to be after the other add subdirs for it to work. As appears maybe the tests aren't importing corrosion but getting it from the other subdirs automatically?

File: CMakeLists.txt [hunk 2]
```diff
@@ -13,6 +13,8 @@ option(USE_QT5 "Use Qt5 even if Qt6 found" OFF)
 option(VCPKG "Use vcpkg for release mode dependencies" OFF)
 option(VCPKG_BUILDTREES_ROOT "Use custom vcpkg buildtrees" "")
 
+option(CXX_QT_BUILD_TESTS "Build the tests for CXX-Qt" ON)
```
Comment (ahayzen-kdab) [inline CMakeLists.txt:16]: do we need the `CXX_QT` prefix ? Could it just be `BUILD_TESTS` as it's kinda common to have `BUILD_TESTS` and `BUILD_EXAMPLES` i thought?

Comment (mattkdab) [reply CMakeLists.txt:16]: i felt it was better to be more descriptive, but maybe it doesnt matter

Comment (ahayzen-kdab) [reply CMakeLists.txt:16]: As we currently have `USE_QT5` and `VCPKG` we probably don't need the prefix, as either they should all have it or they should all not it to be consistent

File: CMakeLists.txt (hunk 1 above)
Comment (mattkdab) [reply CMakeLists.txt:255]: aha!

File: CMakeLists.txt (hunk 2 above)
Comment (redstrate) [reply CMakeLists.txt:16]: If we go with a generic name, make sure it's `BUILD_TESTING` as seems to be more of the standard across CMake projects

Comment (ahayzen-kdab) [reply CMakeLists.txt:16]: Aha thanks, i knew i had seen something like that before :sweat_smile:

File: CMakeLists.txt [hunk 3]
```diff
@@ -219,34 +221,39 @@ endif()
...
+    file(GLOB CXX_QT_GEN_TEST_INPUTS ${CMAKE_CURRENT_SOURCE_DIR}/crates/cxx-qt-gen/test_inputs/*.rs)
+    file(GLOB CXX_QT_GEN_TEST_OUTPUTS ${CMAKE_CURRENT_SOURCE_DIR}/crates/cxx-qt-gen/test_outputs/*.rs)
+    add_test(NAME cxx_qt_gen_test_inputs_gen COMMAND rustfmt --check ${CXX_QT_GEN_TEST_INPUTS})
+    add_test(NAME cxx_qt_gen_test_outputs_gen COMMAND rustfmt --check ${CXX_QT_GEN_TEST_OUTPUTS})
+    
+    # Add test which checks that a build rerun doesn't recompile and uses caches instead
+    add_test(NAME cargo_build_rerun COMMAND "${CMAKE_CURRENT_SOURCE_DIR}/scripts/check_cargo_build_rerun.sh" "${CMAKE_CURRENT_SOURCE_DIR}")
+    
+    # Ensure that cargo_build_rerun doesn't run while we are already building
+    set_tests_properties(cargo_build_rerun PROPERTIES RUN_SERIAL TRUE)
```
Comment (ahayzen-kdab) [inline CMakeLists.txt:256]: can we add these to the preview if block as they are linked to the ` add_test(NAME cargo_clippy` etc some are modifying the env for those and the rest are adding more cargo related tests

Then just leave the add_subdir of tests for this block

File: examples/qml_minimal/CMakeLists.txt [hunk 4]
```diff
@@ -100,47 +100,49 @@ target_link_libraries(${APP_NAME} PRIVATE ${APP_NAME}_lib)
...
+            ${APP_NAME}_${TEST_NAME}_test PROPERTIES ENVIRONMENT_MODIFICATION
+                                                    "${RUNTIME_ENV}"
+        )
+        if (COMMAND add_valgrind_test)
+        add_valgrind_test(
+            ${APP_TEST_NAME} "${TEST_CMD}" ${CMAKE_CURRENT_BINARY_DIR}
+        )
+        else()
+        MESSAGE(STATUS "add_valgrind_test is defined in the top level of CXX-Qt. It will not executed")
+        endif()
```
Comment (ahayzen-kdab) [inline examples/qml_minimal/CMakeLists.txt:144]: looks like the indenting was lost here for the contents of the if branches

Comment (mattkdab) [reply examples/qml_minimal/CMakeLists.txt:144]: oop

File: CMakeLists.txt (hunk 3 above)
Comment (ahayzen-kdab) [reply CMakeLists.txt:250]: i meant the preview `if(BUILD_TESTING)` block :-)  so around line 228

Comment (mattkdab) [reply CMakeLists.txt:250]: oh my

**Review by ahayzen-kdab — APPROVED** [review]
LGTM :-)

**Pull request merged**